### PR TITLE
[EAP7-317] recent configuration changes

### PR DESCRIPTION
**Pull request by honza-kasik**

### CODE REVIEW

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        page.navigate();
+    }
+
+    @AfterClass
+    public static void afterClass() throws IOException, OperationException, TimeoutException, InterruptedException {
+        operations.removeIfExists(serviceAddress);
+        Administration administration = new Administration(client);
+        administration.restartIfRequired();
+        administration.reloadIfRequired();
+    }
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java:94]: What about closing the client?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    @Before
+    public void before() throws IOException, OperationException {
+        //make sure the service is up and running
+        if (operations.invoke("list-changes", serviceAddress).isFailed()) {
+            operations.add(serviceAddress, Values.of("max-history", 100)).assertSuccess();
+        }
+        page.navigate();
+    }
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java:83]: You should make sure you are testing always the same => shouldn't you make sure that the max-history will always be 100 => even when list-changes op didn't fail?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                                    page.getConfigurationChangesTable().getAllConfigurationChanges()));
+
+        page.refresh();
+
+        Assert.assertTrue("Some unexpected differences found! Either some changes are not displayed or some extra appeared.",
+                isChangesListsEqual(changesResource.getAllConfigurationChanges(),
+                                    page.getConfigurationChangesTable().getAllConfigurationChanges()));
+    }
+
+    private boolean isEqualOneChangeToOther(ConfigurationChange a, ConfigurationChange b) {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java]: Maybe better name would be `areChangesEqual` WDYT?

File: common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/EnableConfigurationChangesDialog.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.jboss.hal.testsuite.fragment.runtime.configurationchanges;
+
+import org.jboss.hal.testsuite.fragment.WindowFragment;
+
+/**
+ * @author Jan Kasik <jkasik@redhat.com>
+ *         Created on 9/12/16.
+ */
+public class EnableConfigurationChangesDialog extends WindowFragment {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/EnableConfigurationChangesDialog.java:11]: Javadoc please

File: common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/table/ConfigurationChangesTableFragment.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        return getAllRows().stream()
+                .map((ConfigurationChangeRowFragment row) -> (ConfigurationChange)row)
+                .collect(Collectors.toList());
+    }
+
+    public State getState() {
+        return new State(getAllRows(), getAllConfigurationChanges());
+    }
+
+    public class State implements ConfigurationChangesProvider {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/table/ConfigurationChangesTableFragment.java]: Javadoc please

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/GenericResourceTableFragment.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+import java.lang.reflect.ParameterizedType;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Created by jcechace on 23/02/14.
+ */
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/GenericResourceTableFragment.java:24]: Javadoc fix please. + Copy paste issue :-)

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/GenericResourceTableFragment.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+public class GenericResourceTableFragment<T extends ResourceTableRowFragment> extends BaseFragment {
+
+    // TODO: use properties (this is only temporary implementation)
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/GenericResourceTableFragment.java]: Is the TODO valid? shouldn't be in the PR final implementation?

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import org.jboss.hal.testsuite.util.Console;
+import org.openqa.selenium.By;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.WebElement;
+
+/**
+ * @author Jan Kasik <jkasik@redhat.com>
+ *         Created on 9/12/16.
+ */
+public class ConfigurationChangesPage extends BasePage implements Navigatable {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java:20]: Javadoc please

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    public void refresh() {
+        clickButton("Refresh");
+    }
+
+    public Filter filter() {
+        By filterInputSelector = ByJQuery.selector(".gwt-TextBox");
+        return new Filter(getContentRoot().findElement(filterInputSelector));
+    }
+
+    public class Filter {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java]: Javadoc please. Also if the class can be static, make it static and final.

File: common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/table/ConfigurationChangesTableFragment.java (hunk 5 above)
Comment (rhatlapa) [reply common/src/main/java/org/jboss/hal/testsuite/fragment/runtime/configurationchanges/table/ConfigurationChangesTableFragment.java]: Make the class static and final

File: common/src/main/java/org/jboss/hal/testsuite/util/configurationchanges/ConfigurationChangesResource.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+                        configurationChange.getOperation().contains(filter) ||
+                        configurationChange.getResourceAddress().contains(filter);
+            } catch (Exception e) {
+                e.printStackTrace();
+                return false;
+            }
+        });
+    }
+
+    private final class Change implements ConfigurationChange {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/util/configurationchanges/ConfigurationChangesResource.java]: Make the class static

**Review by rhatlapa — CHANGES_REQUESTED** [review]
Inner classes should be by default static if there is no specific reason to not have them static.
- javadoc improvements
- tests should be failing due reported issues, not having commented out the failing parts.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                        .and("service", "configuration-changes");
+        changesResource = new ConfigurationChangesResource(client);
+    }
+
+    @Before
+    public void before() throws IOException, OperationException {
+        //make sure the service is up and running
+        if (!operations.exists(serviceAddress)) {
+            operations.add(serviceAddress, Values.of("max-history", 100)).assertSuccess();
+        }
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java:81]: Shouldn't be here else branch with write-attribute setting max-history to 100 to have consistent state for all tests?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        Assert.assertTrue("Some unexpected differences found! Either some changes are not displayed or some extra appeared. Some could fail due to the https://issues.jboss.org/browse/HAL-1182",
+                isChangesListsEqual(changesResource.getAllConfigurationChanges(),
+                                    page.getConfigurationChangesTable().getAllConfigurationChanges()));
+    }
+
+    @Test
+    //TODO remove https://issues.jboss.org/browse/HAL-1182 message
+    public void filterConfigurationChangesCapabilities() throws Exception {
+        List<ConfigurationChange> filtered_S_configurationChanges = page.filter().setFilter("s").getAllConfigurationChanges();
+        Assert.assertTrue("Some unexpected differences found! Either some changes are not displayed or some extra appeared. Some could fail due to the https://issues.jboss.org/browse/HAL-1182",
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java]: You could update the sentence to describe the situation in which case it fails due HAL-1182. Also putting the message to some constant might be good idea as you are repeating the same assert message multiple times.

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+    public void refresh() {
+        clickButton("Refresh");
+    }
+
+    public Filter filter() {
+        return new Filter(this);
+    }
+
+    public static final class Filter {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/ConfigurationChangesPage.java:64]: Javadoc please

File: common/src/main/java/org/jboss/hal/testsuite/util/configurationchanges/ConfigurationChangesResource.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+                        configurationChange.getOperation().contains(filter) ||
+                        configurationChange.getResourceAddress().contains(filter);
+            } catch (Exception e) {
+                e.printStackTrace();
+                return false;
+            }
+        });
+    }
+
+    private static final class Change implements ConfigurationChange {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/util/configurationchanges/ConfigurationChangesResource.java:165]: Javadoc might be nice :-)

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        page.navigate();
+    }
+
+    @AfterClass
+    public static void afterClass() throws IOException, OperationException, TimeoutException, InterruptedException {
+        try {
+            operations.removeIfExists(serviceAddress);
+            Administration administration = new Administration(client);
+            administration.restartIfRequired();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java]: Is this really needed, isn't reloadIfRequired satisfactory?

**Review by rhatlapa — CHANGES_REQUESTED** [review]
Rather minor issues. Please update them and you can squash the review fixing commits if you find it suitable.

**Comment by honza-kasik** [comment]
Worked in changes and squashed

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    @Before
+    public void before() throws IOException, OperationException {
+        final int MAX_HISTORY = 100;
+        //make sure the service is up and running
+        if (!operations.exists(serviceAddress)) {
+            operations.add(serviceAddress, Values.of("max-history", MAX_HISTORY)).assertSuccess();
+        } else {
+            operations.writeAttribute(serviceAddress, "max-history", MAX_HISTORY);
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/server/configurationchanges/ConfigurationChangesTestCase.java:80]: You are missing `assertSuccess()` :-)

**Review by rhatlapa — APPROVED** [review]

**Comment by rhatlapa** [comment]
Thanks, there is just minor finding of no big deal. Merging.

**Pull request merged**